### PR TITLE
pyverbs: Add support for null memory regions.

### PR DESCRIPTION
**Pull request by TaranovK**
The patch adds support for null MRs for pyverbs ~~as well as for rdmacm~~.
~~The patch also adds man pages for rdmacm.~~

### CODE REVIEW

**Comment by rleon** [comment]
Why do we need it in rdma-cm?

**Comment by TaranovK** [comment]
> Why do we need it in rdma-cm?

It was a side effect of implementing is_null in pyverbs. Another solution is simply raise an error in pyverb if a client used rdmacm.
To be honest, I never understood why rdmacm implements rdma_reg_* in the first place, and why somebody reused that useless API in pyverbs. 

If you want I can rewrite the commit and remove changes to rdmacm.

**Comment by rleon** [comment]
> > Why do we need it in rdma-cm?
> 
> It was a side effect of implementing is_null in pyverbs. Another solution is simply raise an error in pyverb if a client used rdmacm.
> To be honest, I never understood why rdmacm implements rdma_reg_* in the first place, and why somebody reused that useless API in pyverbs.
> 
> If you want I can rewrite the commit and remove changes to rdmacm.

I'll leave to @EdwardSro to comment about pyverbs, but regarding rdma-cm, please remove.
We don't want to add API that not in use.

Thanks

File: pyverbs/mr.pyx [hunk 1]
```diff
@@ -95,15 +100,19 @@ cdef class MR(PyverbsCM):
             memset(self.buf, 0, length)
         if isinstance(creator, PD):
             pd = <PD>creator
-            if implicit:
+            if null_mr:
+                self.mr = v.ibv_alloc_null_mr(pd.pd)
```
Comment (EdwardSro) [inline pyverbs/mr.pyx:104]: why not to add this check at the beginning (right after line 60), and after successfully allocating a null mr you can just return. IMO it'd be more clear.

File: pyverbs/mr.pyx [hunk 2]
```diff
@@ -95,15 +100,19 @@ cdef class MR(PyverbsCM):
...
                 self.mr = v.ibv_reg_mr(pd.pd, NULL, SIZE_MAX, access)
             else:
                 self.mr = v.ibv_reg_mr(pd.pd, self.buf, length, access)
             self.pd = pd
             pd.add_ref(self)
         elif isinstance(creator, CMID):
             cmid = <CMID>creator
-            if access == e.IBV_ACCESS_LOCAL_WRITE:
+            if null_mr:
+                self.mr = NULL
```
Comment (EdwardSro) [inline pyverbs/mr.pyx:114]: I think it's better to just remove this check, and add an explanation to the docstring, something like "Allocate a null memory region (valid only with PD creator)"

**Review by EdwardSro — CHANGES_REQUESTED** [review]

**Comment by EdwardSro** [comment]
> To be honest, I never understood why rdmacm implements rdma_reg_* in the first place, and why somebody reused that useless API in pyverbs.

They were added to Pyverbs as part of other rdmacm APIs coverage and test flows.
But I agree that they're useless...

**Comment by rleon** [comment]
What is the status of this PR?

Thanks

**Comment by TaranovK** [comment]
Hi,

Our team decided not to continue working on this issue, as we fully switched to C/C++ in our project.  

Best,
Konstantin

**Pull request closed without merging**